File: src/backend/src/pke_backend/crypto/primitives/keywrap.py
"""ECDH(P-256) + HKDF-SHA256 + AES-256-GCM snapshot-key wrap primitive (test-only).

Pins the v0.1 construction from ``context/16_canonical_encoding.md`` §HKDF-SHA256
(snapshot key wrap):

  shared_secret = ECDH(sender_priv, recipient_pub)
  wrapping_key  = HKDF(secret=shared_secret,
                       salt=b"pke/v0.1/keywrap/salt",
                       info=b"pke/v0.1/keywrap/info"
                            || u16be(len(snapshot_id_utf8)) || snapshot_id_utf8
                            || u16be(len(recipient_pub_raw)) || recipient_pub_raw,
                       length=32)
  aad           = b"pke/v0.1/keywrap/aad"
                  || u16be(len(snapshot_id_utf8)) || snapshot_id_utf8
  wrapped       = nonce || AES-256-GCM_Encrypt(key=wrapping_key,
                                               nonce=nonce,
                                               plaintext=snapshot_key,
                                               aad=aad)

Import is restricted to ``src/backend/tests/**`` and ``src/shared/tools/**``
by the ruff banned-api rule and the AST boundary test.
"""

from __future__ import annotations

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec

from pke_backend.crypto.errors import WrapError
from pke_backend.crypto.kdf import hkdf_sha256
from pke_backend.crypto.primitives import aead as _aead

__all__ = ["unwrap", "wrap"]

_HKDF_SALT = b"pke/v0.1/keywrap/salt"
_HKDF_INFO_PREFIX = b"pke/v0.1/keywrap/info"
_AAD_PREFIX = b"pke/v0.1/keywrap/aad"
_WRAPPING_KEY_LEN = 32
_NONCE_LEN = 12
_TAG_LEN = 16
_SNAPSHOT_KEY_LEN = 32
_PUB_RAW_LEN = 65


def _u16be(n: int) -> bytes:
    if n < 0 or n > 0xFFFF:
        raise WrapError(reason=f"u16be input out of range: {n}")
    return n.to_bytes(2, "big")


def _recipient_pub_raw(recipient_public_key: ec.EllipticCurvePublicKey) -> bytes:
    raw = recipient_public_key.public_bytes(
        encoding=serialization.Encoding.X962,
        format=serialization.PublicFormat.UncompressedPoint,
    )
    if len(raw) != _PUB_RAW_LEN:
        raise WrapError(reason=f"recipient public key must be {_PUB_RAW_LEN}-byte uncompressed point, got {len(raw)}")
    return raw


def _check_p256(key: object, name: str) -> None:
    if not isinstance(key, (ec.EllipticCurvePrivateKey, ec.EllipticCurvePublicKey)):
        raise WrapError(reason=f"{name} must be an EC private/public key, got {type(key).__name__}")
    if not isinstance(key.curve, ec.SECP256R1):
        raise WrapError(reason=f"{name} must use P-256, got curve {key.curve.name}")


def _derive_wrapping_key(
    *,
    own_private_key: ec.EllipticCurvePrivateKey,
    peer_public_key: ec.EllipticCurvePublicKey,
    recipient_public_key: ec.EllipticCurvePublicKey,
    snapshot_id: str,
) -> tuple[bytes, bytes]:
    """Return ``(wrapping_key, aad)`` for the v0.1 construction.

    ``own_private_key``/``peer_public_key`` are the two ECDH endpoints from the
    caller's perspective; ``recipient_public_key`` is the recipient regardless
    of which side is calling, because the HKDF ``info`` is always bound to the
    recipient's public key per ``context/16_canonical_encoding.md``.
    """
    if not isinstance(snapshot_id, str) or not snapshot_id:
        raise WrapError(reason="snapshot_id must be a non-empty string")
    snapshot_id_utf8 = snapshot_id.encode("utf-8")
    pub_raw = _recipient_pub_raw(recipient_public_key)

    shared_secret = own_private_key.exchange(ec.ECDH(), peer_public_key)
    info = _HKDF_INFO_PREFIX + _u16be(len(snapshot_id_utf8)) + snapshot_id_utf8 + _u16be(len(pub_raw)) + pub_raw
    wrapping_key = hkdf_sha256(shared_secret, _HKDF_SALT, info, _WRAPPING_KEY_LEN)
    aad = _AAD_PREFIX + _u16be(len(snapshot_id_utf8)) + snapshot_id_utf8
    return wrapping_key, aad


def wrap(
    *,
    sender_private_key: ec.EllipticCurvePrivateKey,
    recipient_public_key: ec.EllipticCurvePublicKey,
    snapshot_id: str,
    snapshot_key: bytes,
    aead_nonce: bytes,
) -> bytes:
    """Return ``nonce || ciphertext || tag`` for the v0.1 snapshot-key wrap."""
    _check_p256(sender_private_key, "sender_private_key")
    _check_p256(recipient_public_key, "recipient_public_key")
    if not isinstance(snapshot_key, (bytes, bytearray)):
        raise WrapError(reason=f"snapshot_key must be bytes, got {type(snapshot_key).__name__}")
    if len(snapshot_key) != _SNAPSHOT_KEY_LEN:
        raise WrapError(reason=f"snapshot_key must be {_SNAPSHOT_KEY_LEN} bytes, got {len(snapshot_key)}")
    if not isinstance(aead_nonce, (bytes, bytearray)):
        raise WrapError(reason=f"aead_nonce must be bytes, got {type(aead_nonce).__name__}")
    if len(aead_nonce) != _NONCE_LEN:
        raise WrapError(reason=f"aead_nonce must be {_NONCE_LEN} bytes, got {len(aead_nonce)}")

    wrapping_key, aad = _derive_wrapping_key(
        own_private_key=sender_private_key,
        peer_public_key=recipient_public_key,
        recipient_public_key=recipient_public_key,
        snapshot_id=snapshot_id,
    )
    ct_and_tag = _aead.encrypt(wrapping_key, bytes(aead_nonce), bytes(snapshot_key), aad)
    return bytes(aead_nonce) + ct_and_tag


def unwrap(
    *,
    recipient_private_key: ec.EllipticCurvePrivateKey,
    sender_public_key: ec.EllipticCurvePublicKey,
    snapshot_id: str,
    wrapped: bytes,
) -> bytes:
    """Recover the 32-byte snapshot key from a ``nonce || ciphertext || tag`` blob."""
    _check_p256(recipient_private_key, "recipient_private_key")
    _check_p256(sender_public_key, "sender_public_key")
    if not isinstance(wrapped, (bytes, bytearray)):
        raise WrapError(reason=f"wrapped must be bytes, got {type(wrapped).__name__}")
    if len(wrapped) < _NONCE_LEN + _TAG_LEN:
        raise WrapError(reason=f"wrapped shorter than nonce+tag: {len(wrapped)} bytes")

    nonce = bytes(wrapped[:_NONCE_LEN])
    ct_and_tag = bytes(wrapped[_NONCE_LEN:])

    wrapping_key, aad = _derive_wrapping_key(
        own_private_key=recipient_private_key,
        peer_public_key=sender_public_key,
        recipient_public_key=recipient_private_key.public_key(),
        snapshot_id=snapshot_id,
    )
    try:
        plaintext = _aead.decrypt(wrapping_key, nonce, ct_and_tag, aad)
    except Exception as exc:
        raise WrapError(reason="AEAD tag verification failed on unwrap") from exc
    if len(plaintext) != _SNAPSHOT_KEY_LEN:
        raise WrapError(reason=f"unwrapped key has wrong length: {len(plaintext)}")
    return plaintext

File: src/ios/PKE/Services/Crypto/KeyWrap.swift
// End-to-end snapshot-key wrap for the PKE v0.1 `ecdhp256+aesgcm256`
// construction. The byte layout is pinned by HLAM-3 /
// `context/16_canonical_encoding.md` §HKDF-SHA256:
//
//   shared_secret = ECDH(P-256)  → 32-byte X coordinate (raw)
//   salt          = b"pke/v0.1/keywrap/salt"
//   info          = b"pke/v0.1/keywrap/info"
//                ‖ u16be(len(snapshot_id_utf8)) ‖ snapshot_id_utf8
//                ‖ u16be(len(recipient_pub_raw)) ‖ recipient_pub_raw  (65 bytes, 0x04‖X‖Y)
//   aad           = b"pke/v0.1/keywrap/aad"
//                ‖ u16be(len(snapshot_id_utf8)) ‖ snapshot_id_utf8
//   wrapping_key  = HKDF-SHA256(shared_secret, salt, info, 32)
//   wrapped       = nonce(12) ‖ AES-256-GCM_encrypt(snapshot_key, wrapping_key, nonce, aad)
//                 = nonce(12) ‖ ciphertext(32) ‖ tag(16)  → 60 bytes total
//
// Any future protocol bump MUST mint a new function alongside this one and a
// new `wrapping_algorithm` identifier; the v0.1 constants and layout are never
// mutated in place. Error reasons reference offsets, lengths, and non-secret
// labels only — never key bytes or plaintext.

import Crypto
import Foundation

public enum KeyWrap {

    // Wire-format constant. Callers may use this when sizing buffers.
    public static let wrappedByteCount = 60

    // Locked v0.1 labels.
    static let saltBytes = Data("pke/v0.1/keywrap/salt".utf8)
    static let infoPrefix = Data("pke/v0.1/keywrap/info".utf8)
    static let aadPrefix = Data("pke/v0.1/keywrap/aad".utf8)

    // Wire-format slice offsets for an unwrapped 60-byte blob.
    private static let nonceRange = 0..<12
    private static let ciphertextRange = 12..<44
    private static let tagRange = 44..<60

    // MARK: - Public API

    /// Wrap `snapshotKey` for `recipientPublic` under the owner's identity,
    /// returning 60 bytes laid out as `nonce ‖ ciphertext ‖ tag`. The nonce is
    /// freshly generated from the platform CSPRNG on every call.
    public static func wrap(
        snapshotKey: SymmetricKey,
        ownerPrivate: P256.KeyAgreement.PrivateKey,
        recipientPublic: AgreementPublicKey,
        snapshotId: String
    ) throws -> Data {
        try wrapWithFixedNonce(
            snapshotKey: snapshotKey,
            ownerPrivate: ownerPrivate,
            recipientPublic: recipientPublic,
            snapshotId: snapshotId,
            nonce: AES.GCM.Nonce()
        )
    }

    /// Unwrap a 60-byte `wrapped` blob from `ownerPublic` to recover the
    /// snapshot key. Any failure (length, tag mismatch, AAD mismatch, point
    /// rejection) surfaces as `CryptoError.aead`. The length gate runs before
    /// any crypto so obviously malformed inputs reject in constant work.
    public static func unwrap(
        _ wrapped: Data,
        recipientPrivate: P256.KeyAgreement.PrivateKey,
        ownerPublic: AgreementPublicKey,
        snapshotId: String
    ) throws -> SymmetricKey {
        guard wrapped.count == wrappedByteCount else {
            throw CryptoError.aead(
                reason: "expected \(wrappedByteCount) bytes, got \(wrapped.count)"
            )
        }
        let nonceData = wrapped.subdata(in: offsetRange(nonceRange, in: wrapped))
        let ctData = wrapped.subdata(in: offsetRange(ciphertextRange, in: wrapped))
        let tagData = wrapped.subdata(in: offsetRange(tagRange, in: wrapped))

        let sharedBytes = try sharedSecretBytes(
            localPrivate: recipientPrivate,
            peerPublic: ownerPublic.underlying
        )
        // `info` is built with the RECIPIENT's pub in both directions; here the
        // recipient is "us", so derive it from our own private key.
        let recipientPubRaw = recipientPrivate.publicKey.x963Representation
        let wrappingKey = try deriveWrappingKey(
            sharedSecret: sharedBytes,
            snapshotId: snapshotId,
            recipientPubRaw: Data(recipientPubRaw)
        )
        let aad = try buildAad(snapshotId: snapshotId)

        let box: AES.GCM.SealedBox
        do {
            let nonce = try AES.GCM.Nonce(data: nonceData)
            box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ctData, tag: tagData)
        } catch {
            throw CryptoError.aead(reason: "wrapped layout rejected by AEAD")
        }
        let plaintext: Data
        do {
            plaintext = try AES.GCM.open(box, using: wrappingKey, authenticating: aad)
        } catch {
            throw CryptoError.aead(reason: "tag verification failed")
        }
        return SymmetricKey(data: plaintext)
    }

    // MARK: - package-visible internals (callable from @testable tests)

    /// Deterministic-nonce variant of `wrap`. The public API never exposes this
    /// surface; it exists so vector tests can pin the nonce supplied by the
    /// fixture and assert byte-for-byte parity against the Python primitive.
    static func wrapWithFixedNonce(
        snapshotKey: SymmetricKey,
        ownerPrivate: P256.KeyAgreement.PrivateKey,
        recipientPublic: AgreementPublicKey,
        snapshotId: String,
        nonce: AES.GCM.Nonce
    ) throws -> Data {
        let sharedBytes = try sharedSecretBytes(
            localPrivate: ownerPrivate,
            peerPublic: recipientPublic.underlying
        )
        let recipientPubRaw = Data(recipientPublic.underlying.x963Representation)
        let wrappingKey = try deriveWrappingKey(
            sharedSecret: sharedBytes,
            snapshotId: snapshotId,
            recipientPubRaw: recipientPubRaw
        )
        let aad = try buildAad(snapshotId: snapshotId)
        let snapshotBytes = snapshotKey.withUnsafeBytes { Data($0) }
        let sealed: AES.GCM.SealedBox
        do {
            sealed = try AES.GCM.seal(
                snapshotBytes,
                using: wrappingKey,
                nonce: nonce,
                authenticating: aad
            )
        } catch {
            throw CryptoError.aead(reason: "AEAD seal failed")
        }
        var out = Data(capacity: wrappedByteCount)
        out.append(contentsOf: Data(nonce))
        out.append(sealed.ciphertext)
        out.append(sealed.tag)
        return out
    }

    static func sharedSecretBytes(
        localPrivate: P256.KeyAgreement.PrivateKey,
        peerPublic: P256.KeyAgreement.PublicKey
    ) throws -> Data {
        // `sharedSecretFromKeyAgreement` only throws on point validation,
        // which CryptoKit performs on key construction — by the time we hold
        // typed keys this call is total. We still surface a typed `.aead`
        // instead of crashing so adversarial input on a future code path
        // cannot turn into a DoS.
        let shared: SharedSecret
        do {
            shared = try localPrivate.sharedSecretFromKeyAgreement(with: peerPublic)
        } catch {
            throw CryptoError.aead(reason: "ecdh failed")
        }
        return shared.withUnsafeBytes { Data($0) }
    }

    static func deriveWrappingKey(
        sharedSecret: Data,
        snapshotId: String,
        recipientPubRaw: Data
    ) throws -> SymmetricKey {
        let info = try buildHkdfInfo(snapshotId: snapshotId, recipientPubRaw: recipientPubRaw)
        return HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: sharedSecret),
            salt: saltBytes,
            info: info,
            outputByteCount: 32
        )
    }

    static func buildHkdfInfo(snapshotId: String, recipientPubRaw: Data) throws -> Data {
        let snapshotIdBytes = Data(snapshotId.utf8)
        var out = Data(capacity: infoPrefix.count + 2 + snapshotIdBytes.count + 2 + recipientPubRaw.count)
        out.append(infoPrefix)
        out.append(try u16be(snapshotIdBytes.count))
        out.append(snapshotIdBytes)
        out.append(try u16be(recipientPubRaw.count))
        out.append(recipientPubRaw)
        return out
    }

    static func buildAad(snapshotId: String) throws -> Data {
        let snapshotIdBytes = Data(snapshotId.utf8)
        var out = Data(capacity: aadPrefix.count + 2 + snapshotIdBytes.count)
        out.append(aadPrefix)
        out.append(try u16be(snapshotIdBytes.count))
        out.append(snapshotIdBytes)
        return out
    }

    static func u16be(_ value: Int) throws -> Data {
        guard value >= 0 && value <= 0xFFFF else {
            throw CryptoError.wrap(reason: "length \(value) does not fit u16be")
        }
        return Data([UInt8(value >> 8), UInt8(value & 0xFF)])
    }

    // MARK: - Helpers

    private static func offsetRange(_ range: Range<Int>, in data: Data) -> Range<Int> {
        let base = data.startIndex
        return (base + range.lowerBound)..<(base + range.upperBound)
    }
}
